lib/Fmt_ast.ml

```diff
@@ -1847,36 +1866,55 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
                        $ fmt "@ ->" )
                    $ fmt "@ "
                    $ cbox 0 (fmt_expression c (sub_exp ~ctx pc_rhs))
-                   $ str ")" $ Cmts.fmt_after c pexp_loc )
-               $ fmt_atrs ))
+                   $ ( match c.conf.indicate_multiline_delimiters with
+                     | `Space -> fits_breaks ")" " )" ~force
+                     | `Closing_on_separate_line ->
+                         fits_breaks ")" ~hint:(1000, 0) ")"
+                     | `No -> str ")" )
+                   $ Cmts.fmt_after c pexp_loc )
+               $ fmt_atrs ) )
       | (lbl, ({pexp_desc= Pexp_function cs; pexp_loc; _} as eN)) :: rev_e1N
         when List.for_all rev_e1N ~f:(fun (_, eI) ->
-                 is_simple c.conf (fun _ -> 0) (sub_exp ~ctx eI)) ->
+                 is_simple c.conf (fun _ -> 0) (sub_exp ~ctx eI) ) ->
           let e1N = List.rev rev_e1N in
           let ctx'' = Exp eN in
           let default_indent = if c.conf.wrap_fun_args then 2 else 4 in
           let indent =
             Params.function_indent c.conf ~ctx ~default:default_indent
           in
           hvbox indent
-            (wrap_if parens "(" ")"
+            (Params.parens_if parens c.conf
                ( hovbox 2
                    (wrap
                       ( fmt_args_grouped e0 e1N $ fmt "@ "
                       $ Cmts.fmt_before c pexp_loc
                       $ fmt_label lbl ":" $ str "(function"
                       $ fmt_attributes c ~pre:Blank ~key:"@"
-                          eN.pexp_attributes ))
-               $ fmt "@ " $ fmt_cases c ctx'' cs $ str ")"
-               $ Cmts.fmt_after c pexp_loc $ fmt_atrs ))
+                          eN.pexp_attributes ) )
+               $ fmt "@ " $ fmt_cases c ctx'' cs
+               $ ( match c.conf.indicate_multiline_delimiters with
+                 | `Space -> fits_breaks ")" " )"
+                 | `Closing_on_separate_line ->
+                     fits_breaks ")" ~hint:(1000, 0) ")"
+                 | `No -> str ")" )
+               $ Cmts.fmt_after c pexp_loc $ fmt_atrs ) )
       | _ ->
           let fmt_atrs =
             fmt_attributes c ~pre:(Break (1, -2)) ~key:"@" pexp_attributes
           in
+          let force =
+            if Location.is_single_line pexp_loc c.conf.margin then Fit
+            else Break
+          in
           fmt_if parens "("
           $ hvbox 2
-              (fmt_args_grouped ~epi:(fmt_atrs $ fmt_if parens ")") e0 e1N1)
-      )
+              ( fmt_args_grouped ~epi:fmt_atrs e0 e1N1
+              $ fmt_if_k parens
+                  ( match c.conf.indicate_multiline_delimiters with
```

Comment:
This pattern is present in many places. Is there a way to abstract it a bit?

Reply:
I factorized in places where it was feasible, in some other places it could still be improved but it does not only depend on `indicate-multiline-delimiters`, parenthesizing as a whole can still be improved